### PR TITLE
SCVMM Plugin, cloud refresh: cacheIpPools

### PR DESCRIPTION
**Pull request by razi-ahmad1**

### CODE REVIEW

File: src/main/groovy/com/morpheus/scvmm/sync/IpPoolsSync.groovy [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+                        poolEnabled  : true,
+                        netmask      : netmask,
+                        subnetAddress: subnetAddress,
+                        type         : poolType,
+                        refType      : 'ComputeZone',
+                        refId        : "${cloud.id}"
+                ]
+                NetworkPool add = new NetworkPool(addConfig)
+                morpheusContext.async.cloud.network.pool.create(add).blockingGet()
+                morpheusContext.async.cloud.network.pool.save(add).blockingGet()
```
Comment (ddeyoung) [inline src/main/groovy/com/morpheus/scvmm/sync/IpPoolsSync.groovy:108]: why save right after a create?

File: src/main/groovy/com/morpheus/scvmm/sync/IpPoolsSync.groovy [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+                    def rangeConfig = [
+                            networkPool: add,
+                            startAddress: it.IPAddressRangeStart,
+                            endAddress: it.IPAddressRangeEnd,
+                            addressCount: (it.TotalAddresses ?: 0).toInteger(),
+                            externalId: it.ID
+                    ]
+                    def newRange = new NetworkPoolRange(rangeConfig)
+                    morpheusContext.async.cloud.network.pool.poolRange.create(newRange).blockingGet()
+                    morpheusContext.async.cloud.network.pool.poolRange.save(newRange).blockingGet()
```
Comment (ddeyoung) [inline src/main/groovy/com/morpheus/scvmm/sync/IpPoolsSync.groovy:120]: why save right after a create?

File: src/main/groovy/com/morpheus/scvmm/sync/IpPoolsSync.groovy [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+            updateList?.each { updateMap ->
+                NetworkPool existingItem = updateMap.existingItem
+                def masterItem = updateMap.masterItem
+                if (existingItem) {
+                    // Update the range (if needed)
+                    if(masterItem.IPAddressRangeStart && masterItem.IPAddressRangeEnd) {
+                        if(!existingItem.ipRanges) {
+                            def range = new NetworkPoolRange(networkPool: existingItem, startAddress: masterItem.IPAddressRangeStart, endAddress: masterItem.IPAddressRangeEnd, addressCount: (masterItem.TotalAddresses ?: 0).toInteger(), externalId: masterItem.ID)
+                            morpheusContext.async.cloud.network.pool.poolRange.save(range).blockingGet()
+                            existingItem.addToIpRanges(range)
```
Comment (ddeyoung) [inline src/main/groovy/com/morpheus/scvmm/sync/IpPoolsSync.groovy:230]: do this before save to reduce the number of api calls.

File: src/main/groovy/com/morpheus/scvmm/sync/IpPoolsSync.groovy [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+                        morpheusContext.async.cloud.network.pool.save(existingItem).blockingGet()
+                    }
+
+                    def existingPermission = morpheusContext.services.resourcePermission.find(new DataQuery()
+                            .withFilter('morpheusResourceType', 'NetworkPool')
+                            .withFilter('morpheusResourceId', existingItem.id)
+                            .withFilter('account', cloud.account))
+                    if(!existingPermission) {
+                        def resourcePerm = new ResourcePermission(morpheusResourceType:'NetworkPool', morpheusResourceId:existingItem.id, account:cloud.account)
+                        morpheusContext.async.resourcePermission.save(resourcePerm).blockingGet()
```
Comment (ddeyoung) [inline src/main/groovy/com/morpheus/scvmm/sync/IpPoolsSync.groovy:297]: shouldn't this be a create and not a save?

**Review by ddeyoung — CHANGES_REQUESTED** [review]
Rely on bulk api methods over single api methods. Most of the save and create usages can be converted to bulkSave and bulkCreate.

File: src/main/groovy/com/morpheus/scvmm/sync/IpPoolsSync.groovy (hunk 2 above)
Comment (razi-ahmad1) [reply src/main/groovy/com/morpheus/scvmm/sync/IpPoolsSync.groovy]: @ddeyoung changes has been done as suggested

File: src/main/groovy/com/morpheus/scvmm/sync/IpPoolsSync.groovy (hunk 3 above)
Comment (razi-ahmad1) [reply src/main/groovy/com/morpheus/scvmm/sync/IpPoolsSync.groovy:250]: @ddeyoung changes has been done as suggested

File: src/main/groovy/com/morpheus/scvmm/sync/IpPoolsSync.groovy (hunk 4 above)
Comment (razi-ahmad1) [reply src/main/groovy/com/morpheus/scvmm/sync/IpPoolsSync.groovy]: @ddeyoung changes has been done as suggested

**Pull request merged**